test/metadata.jl

```diff
@@ -0,0 +1,3065 @@
+    df = DataFrame(a=1:3, b=2:4)
+    leftjoin!(df, df2, on=:a)
+    @test getfield(df, :metadata) === nothing
+    @test isempty(colmetadatakeys(df, :a))
+    @test isempty(colmetadatakeys(df, :b))
+    @test collect(colmetadatakeys(df, :c)) == ["name"]
+    @test colmetadata(df, :c, "name") == "c2"
+
+    df = DataFrame(a=1:3, b=2:4)
+    metadata!(df, "caption", "some1", style=:note)
+    colmetadata!(df, :a, "name", "a1", style=:note)
+    colmetadata!(df, :b, "name", "b1", style=:note)
+    metadata!(df, "caption1", "some1x", style=:none)
+    colmetadata!(df, :a, "name1", "a1x", style=:none)
+    colmetadata!(df, :b, "name1", "b1x", style=:none)
+    leftjoin!(df, df2, on=:a)
+    @test collect(metadatakeys(df)) == ["caption"]
+    @test metadata(df, "caption") == "some1"
+    @test collect(colmetadatakeys(df, :a)) == ["name"]
+    @test colmetadata(df, :a, "name") == "a1"
+    @test collect(colmetadatakeys(df, :b)) == ["name"]
+    @test colmetadata(df, :b, "name") == "b1"
+    @test collect(colmetadatakeys(df, :c)) == ["name"]
+    @test colmetadata(df, :c, "name") == "c2"
+end
+
+@testset "leftjoin" begin
+    for fun in ((x, y) -> leftjoin(x, y, on=:a),
+                (x, y) -> leftjoin(x, y, on=:a, renamecols = "_left" => "_right"))
+        df = DataFrame(a=1:3, b=2:4)
+        df2 = DataFrame(a=1:3, c=1:3)
+
+        res = fun(df, df2)
+        @test getfield(res, :metadata) === nothing
+        @test getfield(res, :colmetadata) === nothing
+
+        metadata!(df2, "caption", "some2", style=:note)
+        colmetadata!(df2, :a, "name", "a2", style=:note)
+        colmetadata!(df2, :c, "name", "c2", style=:note)
+        metadata!(df2, "caption1", "some2x", style=:none)
+        colmetadata!(df2, :a, "name1", "a2x", style=:none)
+        colmetadata!(df2, :c, "name1", "c2x", style=:none)
+        res = fun(df, df2)
+        @test getfield(res, :metadata) === nothing
+        @test isempty(colmetadatakeys(res, 1))
+        @test isempty(colmetadatakeys(res, 2))
+        @test collect(colmetadatakeys(res, 3)) == ["name"]
+        @test colmetadata(res, 3, "name") == "c2"
+
+        metadata!(df, "caption", "some1", style=:note)
+        colmetadata!(df, :a, "name", "a1", style=:note)
+        colmetadata!(df, :b, "name", "b1", style=:note)
+        metadata!(df, "caption1", "some1x", style=:none)
+        colmetadata!(df, :a, "name1", "a1x", style=:none)
+        colmetadata!(df, :b, "name1", "b1x", style=:none)
+        res = fun(df, df2)
+        @test collect(metadatakeys(res)) == ["caption"]
+        @test metadata(res, "caption") == "some1"
+        @test collect(colmetadatakeys(res, 1)) == ["name"]
+        @test colmetadata(res, 1, "name") == "a1"
+        @test collect(colmetadatakeys(res, 2)) == ["name"]
+        @test colmetadata(res, 2, "name") == "b1"
+        @test collect(colmetadatakeys(res, 3)) == ["name"]
+        @test colmetadata(res, 3, "name") == "c2"
+    end
```

Comment:
Maybe test what happens when columns have the same name in the two inputs and `makeunique=true` is passed?

Reply:
added such tests